### PR TITLE
fix tfrecord compress on hdfs

### PR DESCRIPTION
**Pull request by bigbear2017**
Default code doesn't support gzip or other compression type on hdfs. This 
has been fixed in this pull request. Simply using following gzip compression, 
it can reduce 6-7 times less space.
```
    TFRecordFileOutputFormat.setCompressOutput(job, true);
    TFRecordFileOutputFormat.setOutputCompressorClass(job, GzipCodec.class);
```

### CODE REVIEW

**Comment by bigbear2017** [comment]
I signed it!

**Comment by jhseu** [comment]
The underlying TFRecords file format supports compression, and I'm worried that the difference in format here will confuse users. Could we use that instead?

**Comment by sahilbadyal** [comment]
@bigbear2017 Do we have to add options to compress the TFRecord files ?

**Comment by bigbear2017** [comment]
@sahilbadyal  
Just add the following code to your job.
TFRecordFileOutputFormat.setCompressOutput(job, true);
TFRecordFileOutputFormat.setOutputCompressorClass(job, GzipCodec.class);

**Comment by sahilbadyal** [comment]
@bigbear2017 Sorry for not asking the question right. I am using Pyspark

**Comment by fhoering** [comment]
Why is this this PR closed ? It has not been merged.
We have exactly the same issue. We need to write compressed TFRecords from Spark to hdfs. Compressing them manually is yet another step to do. In python TFRecordWriter supports  this with TFRecordCompressionType
Why not reusing the settings from FileOutputFormat as proposed here ? As it uses the Hadoop RecordWriter API it seems the standard way to do.

File: hadoop/src/main/java/org/tensorflow/hadoop/io/TFRecordFileOutputFormat.java [hunk 1]
```diff
@@ -32,20 +32,29 @@
   @Override public RecordWriter<BytesWritable, NullWritable> getRecordWriter(
```
Comment (fhoering) [inline hadoop/src/main/java/org/tensorflow/hadoop/io/TFRecordFileOutputFormat.java]: relevant imports are missing:
import org.apache.hadoop.io.compress.CompressionCodec;
import org.apache.hadoop.io.compress.GzipCodec;
import org.apache.hadoop.util.ReflectionUtils;

import java.io.DataOutputStream;

File: hadoop/src/main/java/org/tensorflow/hadoop/io/TFRecordFileOutputFormat.java [hunk 2]
```diff
@@ -32,20 +32,29 @@
...
+      codec = ReflectionUtils.newInstance(codecClass, conf);
+      extension = codec.getDefaultExtension();
+    }
     int bufferSize = TFRecordIOConf.getBufferSize(conf);
-    final FSDataOutputStream fsdos = fs.create(file, true, bufferSize);
+    Path file = getDefaultWorkFile(context, extension);
+    FileSystem fs = file.getFileSystem(conf);
+    final DataOutputStream fsdos = isCompressed ?
+      new DataOutputStream(codec.createOutputStream(fs.create(file, true, bufferSize))) :
+      fs.create(file, true, bufferSize);
```
Comment (fhoering) [inline hadoop/src/main/java/org/tensorflow/hadoop/io/TFRecordFileOutputFormat.java]: `fs.create(file, true, bufferSize)` could be factorized in a variable

**Comment by fhoering** [comment]
I tested the fix. It is working (with minor changes, need to add the right imports) and can be imported from Python by setting
tf.io.TFRecordCompressionType = tf.io.TFRecordCompressionType.GZIP

Unit tests are still working
Maybe a new one can be added for the gzip case

**Comment by jhseu** [comment]
I'm ok with a pull request that sets the TFRecordCompressionType. I'm just worried about using different compression formats. The TFRecord compression is integrated into the file format, and this PR gzips the file which ends up being a different format. Happy to accept a pull request that can set the option in TFRecords.

**Comment by fhoering** [comment]
@jhseu 

Can you develop
> the TFRecord compression is integrated into the file format

What do you mean by ? 
> and this PR gzips the file which ends up being a different format
GZIP is standardized. So it always will be the GZIP format.

From how I understand it (please correct me if I'm wrong)
- TFRecord is a protobuf with schema Example or SequenceExample
- Protobuf doesn't have integrated gzip compression
- If GZIP compression is integrated in the TfRecordWriter Python API it means that there is (somehow) a c++ Gzip outstream class on top of the encoded protobuf
- Here we are in the Java world, TFRecordWriter is writing the Java protobuf Example/SequenceExample
- We need to somehow gzip the protobuf output stream before writing to disk
- So it ends up applying a Java GzipOutputStream
- This is exactly what is proposed here

About the settings, it end up setting this to the Spark/PYSaprk job which also seems fine to me.
Adding yet another config like tfrecordcompressiontype=Gzip will also be confusing.

```spark.hadoop.mapreduce.output.fileoutputformat.compress:true
spark.hadoop.mapreduce.output.fileoutputformat.compress.codec:
org.apache.hadoop.io.compress.GzipCodec```

At the end on my side I would like to have something that works. So any implementation is fine for me.
As indicated I would also provide a new PR. But what is proposed here (fixing using + units tests) looks fine to me.

**Comment by jhseu** [comment]
TFRecord is a format that is essentially a sequence of records stored as bytes. Those can be anything, but are typically tf.Example protos.

The TFRecord has custom compression settings that are different than just gzipping the file, which you can set in the options:
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/io/TFRecordOptions?hl=en

I'm worried about offering an incompatible, but similar compression setting here. If we could make it compatible, then that's fine.

**Comment by jhseu** [comment]
See the code here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/io/record_writer.cc

**Comment by fhoering** [comment]
In my opinion what is important is that TfRecords that are written here can be read by Pythons TFRecordReader.
And this is the case for GZIP compression. I have tested it.
I could be a first commit. This will already be fine for 90% of the users.

And when someone wants to implement all options of ZLIB he can do this in a second step. I would argue user rather want to use a newer compression lib than doing that.

**Comment by jhseu** [comment]
It would also need to work with tf.data.

**Comment by fhoering** [comment]
So maybe adding a new setting 'tfrecord.compressiontype=Gzip' is better because it restrains the number of possibilities and can be extended to what has been implemented in TfRecordOptions

**Comment by fhoering** [comment]
I tried this:
```
tf.data.TFRecordDataset(files, compression_type="GZIP") 
```
It is working.

As suggested I can provide a new PR with the same code + a unit test.
If this is not enough can you please clearly give a timeline on when you will be implementing this on your side or at least give some clear guidelines on what need to be done or clearly say it will never be done ?

**Comment by jhseu** [comment]
@fhoering 
Hmm, yeah, it seems I misunderstood the format, and that does seem like it should work after reading through the code more closely. Feel free to send a PR, preferably with a unit test if you can get one to work (if not, it's fine).

**Comment by fhoering** [comment]
@jhseu 
I create a new PR https://github.com/tensorflow/ecosystem/pull/125/commits
Please review.

**Comment by bigbear2017** [comment]
hmmm.... I thought I would have a merge in this project. T.T

**Comment by bigbear2017** [comment]
anyway, it is good that this has been fixed. : )

**Comment by jhseu** [comment]
Thanks @bigbear2017 for starting this effort! The test that checked that tf.data reads it correctly alleviated my concern about merging this kind of change in.

**Comment by qiaoguan** [comment]
Execuse me, i user spark to transform the original format to TFrecord format, it works , but the tfrecord file is too large , i want to compress it, what should i do?

import org.tensorflow.hadoop.io.TFRecordFileOutputFormat
import org.tensorflow.hadoop.shaded.protobuf.ByteString
import org.apache.hadoop.io.compress.CompressionCodec
import org.apache.hadoop.io.compress.GzipCodec

val tfrecord_sample_dir = sample_dir + localArgs.yyyyMMdd + "/" + localArgs.loc+"_tfrecord"
FileSystem.get(sc.hadoopConfiguration).delete(new Path(tfrecord_sample_dir), true)
tfrecord_sample.saveAsNewAPIHadoopFileTFRecordFileOutputFormat

where tfrecord_sample is a RDD[(BytesWritable, NullWriteable)] format, BytesWritable is a tfrecord example format。 i see saveAsHadoopFile function provide a way to compress the original data, but i do not know how to use it, anyone can help?
/**

Output the RDD to any Hadoop-supported file system, using a Hadoop OutputFormat class
supporting the key and value types K and V in this RDD. Compress the result with the
supplied codec.
*/
def saveAsHadoopFile[F <: OutputFormat[K, V]](
path: String,
codec: Class[_ <: CompressionCodec])(implicit fm: ClassTag[F]): Unit = self.withScope {
val runtimeClass = fm.runtimeClass
saveAsHadoopFile(path, keyClass, valueClass, runtimeClass.asInstanceOf[Class[F]], codec)
}
i try to use
tfrecord_sample.saveAsHadoopFile[TFRecordFileOutputFormat](tfrecord_sample_dir, GzipCodec)
like this , but it doesn't work

**Comment by nnnnwang** [comment]
hello,when i use follow code to compress the tfrecord as gzip format,it seems doesn't work. There is no different between the output file size whether add the  "option("codec","org.apache.hadoop.io.compress.GzipCodec")" or not 

write.mode("overwrite")
 .format("tfrecords")
  .option("recordType","Example")
  .option("codec", "org.apache.hadoop.io.compress.GzipCodec")
  .save(outputFilePath)

**Pull request closed without merging**